[WK1] WebKit XML parsing can deny external entity loads from other in-process libxml2 clients

ddkilzer opened the pull request:
#### 7b1fb05b974f0c50874da628b5388cca5b7292e2
<pre>
[WK1] WebKit XML parsing can deny external entity loads from other in-process libxml2 clients
<a href="https://bugs.webkit.org/show_bug.cgi?id=273045">https://bugs.webkit.org/show_bug.cgi?id=273045</a>
&lt;<a href="https://rdar.apple.com/126476952">rdar://126476952</a>&gt;

Reviewed by Alex Christensen and Michael Catanzaro.

The fix for Bug 259235 replaced the default libxml2 external entity
loader function with one from WebKit that implements a same-origin
policy for the web, but that means that WebKit1 clients that use libxml2
for parsing independent of WebKit also start using this function, which
can cause external entity load failures depending on the libxml2 API
used.

Fix this by setting the external entity loader using
XMLDocumentParserScope, then unsetting it when that object is
deallocated.

Add two more places where XMLDocumentParserScope was missing in
WebCore::XMLDocumentParser::appendFragmentSource() and
WebCore::parseAttributes().

Covered by these tests (among others):
    fast/xsl/xslt-bad-import-uri.html
    http/tests/misc/xslt-bad-import.html
    http/tests/security/contentSecurityPolicy/xsl-redirect-blocked.html
    http/tests/security/cross-origin-xsl-redirect-BLOCKED.html
    http/tests/security/xss-ALLOWED-xsl-external-entity-xslt-docloader.html
    http/tests/security/xss-DENIED-xsl-external-entity-xslt-docloader.html

* Source/WebCore/dom/ProcessingInstruction.cpp:
(WebCore::ProcessingInstruction::checkStyleSheet):
- Move Ref variable for Document to top of method instead of calling
  document() repeatedly.
- Pass extra argument added to WebCore::parseAttributes().

* Source/WebCore/xml/XSLTProcessorLibxslt.cpp:
(WebCore::docLoaderFunc):
- Minor clean-up to inline return statement.
* Source/WebCore/xml/parser/XMLDocumentParser.h:
(WebCore::XMLParserContext::XMLParserContext): Remove.
(WebCore::XMLDocumentParser::XMLDocumentParser): Remove.
- Delete default constructors.
(WebCore::externalEntityLoader): Add declaration.
(WebCore::parseAttributes):
- Add WebCore::CachedResourceLoader to argument list.
* Source/WebCore/xml/parser/XMLDocumentParserLibxml2.cpp:
(WebCore::externalEntityLoader):
- Remove &apos;static&apos; keyword as this function is referenced outside this
  source file.
(WebCore::initializeXMLParser):
- Don&apos;t set external entity loader here since it will be set by
  XMLDocumentParserScope instead.
- Add RELEASE_ASSERT() that the external entity loader isn&apos;t already set
  to WebCore::externalEntityLoader.  This indicates that an
  XMLDocumentParserScope was created too early, and that external entity
  loading will result in a stack recursion crash later.
(WebCore::XMLDocumentParser::initializeParserContext):
- Remove unneeded XMLDocumentParserScope object.
(WebCore::XMLDocumentParser::appendFragmentSource):
- Add missing XMLDocumentParserScope object.
(WebCore::parseAttributes):
- Add WebCore::CachedResourceLoader to argument list so that an
  XMLDocumentParserScope object can be created before calling libxml2 to
  parse XML content.
* Source/WebCore/xml/parser/XMLDocumentParserScope.cpp:
(WebCore::XMLDocumentParserScope::XMLDocumentParserScope):
(WebCore::XMLDocumentParserScope::~XMLDocumentParserScope):
- Update to save the current external entity loader and set WebKit&apos;s
  external entity loader function.
* Source/WebCore/xml/parser/XMLDocumentParserScope.h:
(WebCore::XMLDocumentParserScope::m_oldEntityLoader): Add.
- Add instance variable for saving and restoring the external entity
  loader.

Canonical link: <a href="https://commits.webkit.org/278168@main">https://commits.webkit.org/278168@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2dbf864dfc1421e43a15994f97cc43d24dfdf72e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/49480 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/28762 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/52518 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/52719 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/153 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/51784 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/34780 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/26382 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/40391 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/51580 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/26304 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/42629 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/21509 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/23764 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/43812 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/7849 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/45678 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/44318 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/54231 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/24563 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/20746 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/47757 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/25834 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/42817 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/46773 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/26674 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7151 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/25557 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/WebCore/xml/parser/XMLDocumentParser.h:
```diff
@@ -54,10 +54,7 @@ class XMLParserContext : public RefCounted<XMLParserContext> {
-        : m_context(context)
-    {
-    }
+    XMLParserContext(xmlParserCtxtPtr);
```
```suggestion
    explicit XMLParserContext(xmlParserCtxtPtr);
```

ddkilzer replied:
Fixed.

mcatanzaro reviewed (commented):
Hm, I'm not familiar with libxml2, but I have a feeling that libraries shouldn't be replacing the default entity loader function in the first place... right? It relies on the assumption that no other code outside WebKit will try to do the same. But I took a quick look at the [API docs](https://gnome.pages.gitlab.gnome.org/libxml2/devhelp/libxml2-parser.html) and it looks like there's no way to specify the entity loader function to use without changing the default one, so maybe there's really no other way? And it probably only matters for WebKitLegacy, so the impact is limited. And I do like your change here.

Would a libxml2 bug report be appropriate, to request a way for libraries to safely use their own entity loaders without messing with the application?

mcatanzaro commented on Source/WebCore/xml/parser/XMLDocumentParser.h:
```diff
@@ -190,6 +191,8 @@ class XMLDocumentParser final : public ScriptableDocumentParser, public PendingS
 #endif
 
-std::optional<HashMap<String, String>> parseAttributes(const String&);
+xmlParserInputPtr entityLoader(const char* url, const char* id, xmlParserCtxtPtr);
```
`WebCore::entityLoader` is a very generic name. I would move this into the class scope and make it static, so it can be `WebCore::XMLDocumentParser::entityLoader` instead.

ddkilzer replied:
This function needs to keep its C-linkage since it's called as a function pointer from libxml2.

Can we change it to `WebCore::externalEntityLoader`?

I don't know whether it will keep its C-linkage as a C++ static class method.

mcatanzaro replied:
Are you sure C linkage actually matters? We pass C++ lambda functions to C libraries all the time; the only restriction is no variables can be captured (as then it wouldn't cast to a function pointer). I don't think it can have C linkage if it's part of the WebCore namespace anyway; surely you would need to move it outside the namespace and add an `extern "C"` block?

But anyway, this was a nitpick suggestion. It doesn't really matter. Feel free to do whatever you prefer.

mcatanzaro reviewed (approved):
OK, smart. Feels more robust than the original version of this pull request.

achristensen07 reviewed (approved):
There aren't any tests with WebKit1 clients and non-WebKit libxml clients to verify this change.

Commit:
[WK1] WebKit XML parsing can deny external entity loads from other in-process libxml2 clients
https://bugs.webkit.org/show_bug.cgi?id=273045
<rdar://126476952>

Reviewed by Alex Christensen and Michael Catanzaro.

The fix for Bug 259235 replaced the default libxml2 external entity
loader function with one from WebKit that implements a same-origin
policy for the web, but that means that WebKit1 clients that use libxml2
for parsing independent of WebKit also start using this function, which
can cause external entity load failures depending on the libxml2 API
used.

Fix this by setting the external entity loader using
XMLDocumentParserScope, then unsetting it when that object is
deallocated.

Add two more places where XMLDocumentParserScope was missing in
WebCore::XMLDocumentParser::appendFragmentSource() and
WebCore::parseAttributes().

Covered by these tests (among others):
    fast/xsl/xslt-bad-import-uri.html
    http/tests/misc/xslt-bad-import.html
    http/tests/security/contentSecurityPolicy/xsl-redirect-blocked.html
    http/tests/security/cross-origin-xsl-redirect-BLOCKED.html
    http/tests/security/xss-ALLOWED-xsl-external-entity-xslt-docloader.html
    http/tests/security/xss-DENIED-xsl-external-entity-xslt-docloader.html

* Source/WebCore/dom/ProcessingInstruction.cpp:
(WebCore::ProcessingInstruction::checkStyleSheet):
- Move Ref variable for Document to top of method instead of calling
  document() repeatedly.
- Pass extra argument added to WebCore::parseAttributes().

* Source/WebCore/xml/XSLTProcessorLibxslt.cpp:
(WebCore::docLoaderFunc):
- Minor clean-up to inline return statement.
* Source/WebCore/xml/parser/XMLDocumentParser.h:
(WebCore::XMLParserContext::XMLParserContext): Remove.
(WebCore::XMLDocumentParser::XMLDocumentParser): Remove.
- Delete default constructors.
(WebCore::externalEntityLoader): Add declaration.
(WebCore::parseAttributes):
- Add WebCore::CachedResourceLoader to argument list.
* Source/WebCore/xml/parser/XMLDocumentParserLibxml2.cpp:
(WebCore::externalEntityLoader):
- Remove 'static' keyword as this function is referenced outside this
  source file.
(WebCore::initializeXMLParser):
- Don't set external entity loader here since it will be set by
  XMLDocumentParserScope instead.
- Add RELEASE_ASSERT() that the external entity loader isn't already set
  to WebCore::externalEntityLoader.  This indicates that an
  XMLDocumentParserScope was created too early, and that external entity
  loading will result in a stack recursion crash later.
(WebCore::XMLDocumentParser::initializeParserContext):
- Remove unneeded XMLDocumentParserScope object.
(WebCore::XMLDocumentParser::appendFragmentSource):
- Add missing XMLDocumentParserScope object.
(WebCore::parseAttributes):
- Add WebCore::CachedResourceLoader to argument list so that an
  XMLDocumentParserScope object can be created before calling libxml2 to
  parse XML content.
* Source/WebCore/xml/parser/XMLDocumentParserScope.cpp:
(WebCore::XMLDocumentParserScope::XMLDocumentParserScope):
(WebCore::XMLDocumentParserScope::~XMLDocumentParserScope):
- Update to save the current external entity loader and set WebKit's
  external entity loader function.
* Source/WebCore/xml/parser/XMLDocumentParserScope.h:
(WebCore::XMLDocumentParserScope::m_oldEntityLoader): Add.
- Add instance variable for saving and restoring the external entity
  loader.

Canonical link: https://commits.webkit.org/278168@main
